src/v/storage/disk_log_impl.cc

```diff
@@ -386,15 +388,41 @@ ss::future<> disk_log_impl::compact_adjacent_segments(
         co_await ss::remove_file(compact_index.string());
     }
 
+    // lock the range. only metadata (e.g. open/rename/delete) i/o occurs with
+    // these locks held so it is a relatively short duration. all of the data
+    // copying and compaction i/o occurred above with no locks held.
+    std::vector<ss::rwlock::holder> locks;
+    locks.reserve(segments.size());
+    for (auto& segment : segments) {
+        locks.push_back(co_await segment->write_lock());
```

Comment:
@dotnwat  the for-loop makes them sequential. I think we should dispatch the write-lock acquisition concurrently.

Reply:
deadlocks? currently we acquire locks in < order

Reply:
Ah I see. Good point. I think we should have then a timeout as part of the interface. I think I passed an abort src and a timeout on the up call of compact ()

Reply:
sure. if we have a timeout we can do parallel acquisition. i'll see what i can do